### PR TITLE
cache param env canonicalization

### PR DESCRIPTION
**Pull request by aliemjay**
Canonicalize ParamEnv only once and store it. Then whenever we try to canonicalize `ParamEnvAnd<'tcx, T>` we only have to canonicalize `T` and then merge the results.

Prelimiary results show ~3-4% savings in diesel and serde benchmarks.

Best to review commits individually. Some commits have a short description.

Initial implementation had a soundness bug (https://github.com/rust-lang/rust/pull/117749#issuecomment-1840453387) due to cache invalidation:
- When canonicalizing `Ty<'?0>` we first try to resolve region variables in the current InferCtxt which may have a constraint `?0 == 'static`. This means that we register `Ty<'?0> => Canonical<Ty<'static>>` in the cache, which is obviously incorrect in another inference context.
- This is fixed by not doing region resolution when canonicalizing the query *input* (vs. response), which is the only place where ParamEnv is used, and then in a later commit we *statically* guard against any form of inference variable resolution of the cached canonical ParamEnv's.

r? @ghost

### CODE REVIEW

**Comment by aliemjay** [comment]
> The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7092326727/job/19303382247) [(plain)](https://github.com/rust-lang/rust/commit/bcbfbe62ff1ca24302e585679473c43f59054717/checks/19303382247/logs)

MCVE:
```rust
struct VarZeroVec<'a, F>(&'a (), F);

trait Yokeable<'a> {
    type Output;
}

impl<'a> Yokeable<'a> for () {
    type Output = VarZeroVec<'a, ()>;
}

fn transform_mut<'a, F>(f: F)
where
    F: FnOnce(VarZeroVec<'a, ()>),
{
    f(absurd::<<() as Yokeable<'a>>::Output>())
}

fn absurd<T>() -> T{ todo!() }
```

**Comment by aliemjay** [comment]
changed to fix the small regressions
@bors try @rust-timer queue

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs [hunk 1]
```diff
@@ -126,19 +126,52 @@ impl<'tcx> InferCtxt<'tcx> {
...
+        let (param_env, value) = value.into_parts();
+        let base = self.tcx.canonical_param_env_cache.get_or_insert(
+            param_env,
+            query_state,
+            |query_state| {
+                Canonicalizer::canonicalize(
+                    param_env,
+                    self,
+                    self.tcx,
+                    &CanonicalizeFreeRegionsOtherThanStatic,
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:138]: This does not spark joy 😁 Does anything break if you change this to `CanonicalizeFreeRegions`? I want to get away from canonicalizing static completely. iirc the last time I tried after #102472 there was still some failure in project or sth 🤔 would prefer for you (or me) to first try to  also canonicalize `'static` again to check why it matters.

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs [hunk 2]
```diff
@@ -848,14 +846,8 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
...
     fn canonicalize_ty_var(&mut self, info: CanonicalVarInfo<'tcx>, ty_var: Ty<'tcx>) -> Ty<'tcx> {
-        let infcx = self.infcx;
-        let bound_to = infcx.shallow_resolve(ty_var);
-        if bound_to != ty_var {
-            self.fold_ty(bound_to)
-        } else {
-            let var = self.canonical_var(info, ty_var.into());
-            Ty::new_bound(self.tcx, self.binder_index, var.into())
-        }
+        let var = self.canonical_var(info, ty_var.into());
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:802]: can you debug assert that `ty_var == infcx.shallow_resolve(ty_var)`?

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs [hunk 3]
```diff
@@ -867,13 +859,7 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
...
     ) -> ty::Const<'tcx> {
-        let infcx = self.infcx;
-        let bound_to = infcx.shallow_resolve(const_var);
-        if bound_to != const_var {
-            self.fold_const(bound_to)
-        } else {
-            let var = self.canonical_var(info, const_var.into());
-            ty::Const::new_bound(self.tcx, self.binder_index, var, self.fold_ty(const_var.ty()))
-        }
+        let var = self.canonical_var(info, const_var.into());
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:815]: same here

**Review by lcnr — COMMENTED** [review]
some nits, then r=me, afaict CI was an unrelated issue

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs (hunk 1 above)
Comment (aliemjay) [reply compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:138]: Borrowck error in [`tests/ui/trivial-bounds/trivial-bounds-object.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/trivial-bounds/trivial-bounds-object.rs)

ICE in [`tests/ui/nll/issue-61311-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61311-normalize.rs) and [`tests/ui/nll/issue-61320-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61320-normalize.rs). MCVE:
```rust
pub trait HasProj {
    type Proj;
}

impl<T: ?Sized> HasProj for T {
    type Proj = ();
}

fn test()
where
    dyn Send + 'static: HasProj,
    <dyn Send + 'static as HasProj>::Proj: Clone,
{
    let _ = test;
}
```

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs (hunk 3 above)
Comment (aliemjay) [reply compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:819]: done

**Comment by aliemjay** [comment]
> some nits, then r=me, afaict CI was an unrelated issue

I turns out it is related. Rustdoc incorrectly uses ParamEnv's with inference variables in them. This causes an ICE during canonicalization in one of the `unwrap`s in the commit "make infcx optional in canonicalizer". This was fixed by the last commit by using an empty param env. I assume it does not need a re-approval.

@bors r=lcnr rollup=never

File: compiler/rustc_infer/src/infer/canonical/canonicalizer.rs (hunk 1 above)
Comment (aliemjay) [reply compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:138]: cc #118965. It is also the root cause behind the ICE.

**Pull request merged**